.github/workflows/reusable_publish_version.yml

```diff
@@ -38,12 +37,121 @@ jobs:
+        run: |
+          set -x
+
+          currentVersion=$(/home/runner/.cargo/bin/taplo get -f lib/Cargo.toml "package.version")
+
+          if [[ $currentVersion == *"-beta"* ]]; then
+            major=$(echo $currentVersion | tr "." "\n" | sed -n 1p)
+            minor=$(echo $currentVersion | tr "." "\n" | sed -n 2p)
+            patchAndMeta=$(echo $currentVersion | tr "." "\n" | sed -n 3p)
+            betaNum=$(echo $currentVersion | tr "." "\n" | sed -n 4p)
+            betaVersion=${major}.${minor}.${patchAndMeta}.$(($betaNum + 1))
+          else
+            git checkout -b releases/beta
+            major=$(echo $currentVersion | tr "." "\n" | sed -n 1p)
+            minor=$(echo $currentVersion | tr "." "\n" | sed -n 2p)
+            betaVersion=${major}.$(($minor + 1)).0-beta.1
+          fi
+
+          # Bump the crate version
+          sed -i "s#^version = \".*\"#version = \"${betaVersion}\"#" Cargo.toml
+          sed -i "s#^version = \".*\"#version = \"${betaVersion}\"#" lib/Cargo.toml
+
+          # Update Cargo.lock without updating dependency versions
+          cargo check --no-default-features --features storage-mem
+
+          # Commit changes
+          git commit -am "Prepare v${betaVersion} release"
+
+          # Create the tag
+          git tag -a v${betaVersion} -m "Release ${betaVersion}"
+
+      - name: Push changes
+        if: ${{ inputs.publish && (inputs.environment == 'beta' || inputs.environment == 'release') }}
+        run: git push
+
+      - name: Push tag
+        if: ${{ inputs.publish && (inputs.environment == 'beta' || inputs.environment == 'release') }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.RELEASE_PLZ_TOKEN }} # Need the custom user token here so we can push the tag
+        run: git push --tags
+
       - name: Set outputs
         id: outputs
         run: |
-          echo "git-ref=${{ inputs.git-ref || github.ref_name }}" >> $GITHUB_OUTPUT
-          # If the name is not provided, use the git_ref (i.e. v1.0.0)
-          echo "name=${{ inputs.name || inputs.git-ref || github.ref_name }}" >> $GITHUB_OUTPUT
+          set -x
+
+          if [[ "${{ inputs.environment }}" == "beta" && "${{ inputs.publish }}" == "true" ]]; then
+            echo "git-ref=releases/beta" >> $GITHUB_OUTPUT
+          else
+            echo "git-ref=${{ inputs.git-ref || github.ref_name }}" >> $GITHUB_OUTPUT
+          fi
+
+          if [[ "${{ inputs.environment }}" == "nightly" ]]; then
+            echo "name=${{ inputs.environment }}" >> $GITHUB_OUTPUT
+
+            date=$(git show --no-patch --format=%ad --date=format:%Y%m%d)
+            rev=$(git rev-parse --short HEAD)
+            echo "build-metadata=${date}.${rev}" >> $GITHUB_OUTPUT
+          else
+            version=$(/home/runner/.cargo/bin/taplo get -f lib/Cargo.toml "package.version")
+            echo "name=v${version}" >> $GITHUB_OUTPUT
+
+            echo "build-metadata=\"\"" >> $GITHUB_OUTPUT
+          fi
```

Comment:
Are these steps idempotent? let's say the tests or the build jobs fail, can we trigger this workflow again?